File: content/problems/1964-find-the-longest-valid-obstacle-course-at-each-position/solutions/upper-bound-patience/Solution.java
class Solution {
    public int[] longestObstacleCourseAtEachPosition(int[] obstacles){
        int[] tails=new int[obstacles.length],answer=new int[obstacles.length];
        int size=0;
        for(int i=0;i<obstacles.length;i++){
            int left=0,right=size;
            while(left<right){
                int middle=(left+right)/2;
                if(tails[middle]<=obstacles[i])left=middle+1;
                else right=middle;
            }
            tails[left]=obstacles[i];
            if(left==size)size++;
            answer[i]=left+1;
        }
        return answer;
    }
}

File: content/problems/1964-find-the-longest-valid-obstacle-course-at-each-position/solutions/upper-bound-patience/Solution.kt
class Solution {
    fun longestObstacleCourseAtEachPosition(obstacles:IntArray):IntArray{
        val tails=IntArray(obstacles.size);
        val answer=IntArray(obstacles.size);
        var size=0;
        for(i in obstacles.indices){
            var left=0;
            var right=size;
            while(left<right){
                val middle=(left+right)/2;
                if(tails[middle]<=obstacles[i])left=middle+1 else right=middle
            };
            tails[left]=obstacles[i];
            if(left==size)size++;
            answer[i]=left+1
        };
        return answer
    }
}
